src/ocean/task/TaskPool.d

```diff
@@ -227,30 +187,28 @@ class TaskPool ( TaskT : Task ) : ObjectPool!(Task)
             if (this.num_busy() >= this.limit())
                 return false;
 
-            auto task = cast(TaskT) this.get(new OwnedTask);
+            auto task = cast(TaskT) this.get(new TaskT);
             assert (task !is null);
 
-            try
-            {
-                task.deserialize(serialized);
-                theScheduler.schedule(task);
-            }
-            catch (TaskKillException e)
-            {
-                // don't try recycling task upon TaskKillException as this is not
-                // normal code flow and it may have already been recycled by
-                // finishing on its own
-                throw e;
-            }
-            catch (Exception e)
-            {
-                this.recycle(task);
-                throw e;
-            }
+            task.deserialize(serialized);
+            task.terminationHookUnique(&this.taskTerminationHook);
+            theScheduler.schedule(task);
 
             return true;
         }
     }
+
+    /***************************************************************************
+
+        Used to recycle pool tasks when they finish
+
+    ***************************************************************************/
+
+    private void taskTerminationHook ( )
+    {
+        auto task = Task.getThis();
+        this.recycle(task);
```

Comment:
Good catch!

Reply:
Actually no, this is a wrong lead - `recycle` method will be called anyway after termination hook because of https://github.com/sociomantic-tsunami/ocean/blob/v4.x.x/src/ocean/task/Task.d#L493

But it does show that this PR changes order of recycling calls - not sure yet if it is of any relevance